### ISSUE TITLE
how to avoid race condition on save_report

### ISSUE THREAD

**Issue by danmayer** [danmayer/coverband#303]
This wasn't thought to be much of a problem, but on higher volume production systems the race condition can cause coverage data to be lost. NOTE: this issue doesn't occur until at fairly notable scale.

https://github.com/danmayer/coverband/blob/master/lib/coverband/adapters/base.rb#L53

Basically, if you have dozens of servers with hundreds of total processes, if two processes try to merge at the same time one will win and the other will have it's data lost.

Moving to a write-only and avoid any read or round trip is probably best, but might make deployment significantly more complicated.

**Comment by kbaum** [comment]
Started converting some of the work i did on multi_redis_store to use hincrby

https://github.com/danmayer/coverband/compare/replace_redis_store_with_hash

A few end to end tests failing for some reason still.

**Comment by kbaum** [comment]
Good news is all tests are passing with new HashRedisStore.

Bad news is that execution time has gotten much worse:

With existing redis store:
```
➜  coverband git:(replace_redis_store_with_hash) bundle exec rake benchmarks:redis_reporting
runs benchmarks on reporting large sets of files to redis
Warming up --------------------------------------
   store_reports_all     2.000  i/100ms
Calculating -------------------------------------
   store_reports_all     28.873  (± 6.9%) i/s -    432.000  in  15.018180s
Warming up --------------------------------------
store_reports_subset     4.000  i/100ms
Calculating -------------------------------------
store_reports_subset     42.374  (± 7.1%) i/s -    844.000  in  20.000767s
```

New HashRedisStore:
```
➜  coverband git:(replace_redis_store_with_hash) ✗ bundle exec rake benchmarks:redis_reporting
runs benchmarks on reporting large sets of files to redis
Warming up --------------------------------------
   store_reports_all     1.000  i/100ms
Calculating -------------------------------------
   store_reports_all      0.318  (± 0.0%) i/s -      5.000  in  15.753592s
Warming up --------------------------------------
store_reports_subset     1.000  i/100ms
Calculating -------------------------------------
store_reports_subset     10.624  (± 9.4%) i/s -    211.000  in  20.063152s
```

**Comment by danmayer** [comment]
hmmm OK, so we might need to look at this change in combination with some other ideas...

moving to sparse arrays of hits only for example... Then not reporting any files with no change, etc... In some sense with it in the background we are less concerned with it's overall speed, it was important to be really fast when it was in the request cycle of rack middleware.

**Comment by kbaum** [comment]
We could play with some other formats for storing counters. Currently we are using a separate hash for reach file. Hash in redis has a limitation of only being able to set/incr values in one hash per redis call. If we either moved to one hash for all files or storing file data across separate redis key/values (no hash at all) then we can make all updates in one incr and one set call. The downside of one hash for all files is that I believe hash only supports one expiration for entire hash. The hash could grow large over time without expirations.

> moving to sparse arrays of hits only for example...

What does sparse arrays mean? Like not writing nils and 0 for coverage data?

> Then not reporting any files with no change, etc

I believe the currently implementation only reports on files with changes.

> In some sense with it in the background we are less concerned with it's overall speed

Agreed but reporting also effects startup time of rails and also happens on exit. Best if we can keep it quick.

**Comment by danmayer** [comment]
> What does sparse arrays mean? Like not writing nils and 0 for coverage data?

yeah so like the new format internal to oneshot coverage it is an array `[1,5,12,13]` which stands for those lines, all the missing numbers aren't included at all. Easy for binary as an array... For hit count...something like below... with no keys related to 0s

`{1 => 1, 5 => 9, 12 => 9, 13 => 9}`

> I believe the currently implementation only reports on files with changes.

so I think we did that for awhile and we likely skip some steps like merge on files without changes... but in the end we redis write a single blob at the moment so the means the blob always gets written with a bunch of data that doesn't change... in a key per file we would want to ensure we don't do an update across all keys only keys with changes... I think the filtering related to this might have been lost in various format changes when moving from request storage to background storage.

> Agreed but reporting also effects startup time of rails and also happens on exit. Best if we can keep it quick.

Always true ;)

**Comment by kbaum** [comment]
> For hit count...something like below... with no keys related to 0s

I like this idea but how do we go about differentiating 0's from nils when generating coverage report? I had previously been skipping 0's and nils in the HashRedisStore implementation but realized that I was missing some data when generating the coverage report. For example, my coverage data was missing nil for empty lines at the end of the file that are unreachable.

> so I think we did that for awhile and we likely skip some steps like merge on files without changes...

Was under the impression we are still only reporting differences as we use the Delta class to calculate differences before calling save_coverage on the store.

https://github.com/danmayer/coverband/blob/master/lib/coverband/collectors/coverage.rb#L54

Just to make sure I added a debug message locally to HashRedisStore#save_report

```ruby
      def save_report(report)
        @logger.debug("save_report(#{report.inspect})")
 ```

On a quiet rails instance, I see this in the log every 30 seconds:

```
save_report({})
save_report({})
```

Do you think there is still more to optimize here?

**Comment by kbaum** [comment]
Seems to be a growth in both total allocated memory and total retained. That is surprising to me. Seems to be coming from redis library in pipeline.

#### RedisStore
```
Total allocated: 110796696 bytes (821919 objects)
Total retained:  666256 bytes (4015 objects)

allocated memory by gem
-----------------------------------
  49874880  coverband/lib
  30916130  json
  20644776  redis-4.1.0
   9360910  other

allocated memory by file
-----------------------------------
  30916130  /Users/karlbaum/.rubies/ruby-2.6.3/lib/ruby/2.6.0/json/common.rb
  24716500  /Users/karlbaum/workspace/coverband/lib/coverband/adapters/redis_store.rb
  20069040  /Users/karlbaum/workspace/coverband/lib/coverband/collectors/delta.rb
  11499056  /Users/karlbaum/.gem/ruby/2.6.3/gems/redis-4.1.0/lib/redis/connection/ruby.rb
   9200910  <internal:prelude>
   9141000  /Users/karlbaum/.gem/ruby/2.6.3/gems/redis-4.1.0/lib/redis/connection/command_helper.rb
   5089340  /Users/karlbaum/workspace/coverband/lib/coverband/adapters/base.rb
      3920  /Users/karlbaum/.gem/ruby/2.6.3/gems/redis-4.1.0/lib/redis.rb
       800  /Users/karlbaum/.gem/ruby/2.6.3/gems/redis-4.1.0/lib/redis/client.rb
```

#### New HashRedisStore

```
Total allocated: 300407904 bytes (3746151 objects)
Total retained:  666256 bytes (4015 objects)

allocated memory by gem
-----------------------------------
 274018049  redis-4.1.0
  25579140  coverband/lib
    810715  other

allocated memory by file
-----------------------------------
 145019320  /Users/karlbaum/.gem/ruby/2.6.3/gems/redis-4.1.0/lib/redis/connection/command_helper.rb
 105183449  /Users/karlbaum/.gem/ruby/2.6.3/gems/redis-4.1.0/lib/redis/connection/ruby.rb
  20069040  /Users/karlbaum/workspace/coverband/lib/coverband/collectors/delta.rb
  13056800  /Users/karlbaum/.gem/ruby/2.6.3/gems/redis-4.1.0/lib/redis.rb
   9606000  /Users/karlbaum/.gem/ruby/2.6.3/gems/redis-4.1.0/lib/redis/pipeline.rb
   5510100  /Users/karlbaum/workspace/coverband/lib/coverband/adapters/hash_redis_store.rb
   1152480  /Users/karlbaum/.gem/ruby/2.6.3/gems/redis-4.1.0/lib/redis/client.rb
    650715  <internal:prelude>
```

**Comment by kbaum** [comment]
I've made some improvements to this implementation using a lua script to save all attributes of the hash in one call:
https://github.com/danmayer/coverband/pull/304/files#diff-d3e3760bed63f2ad73ff3b44dd78f8b9R39

Total memory allocated has gone down a great deal compared our current RedisStore implementation. From 110_796_696 to 53_812_980.
```
Total allocated: 53_812_980 bytes (684850 objects)
 47 Total retained:  666256 bytes (4015 objects)
 48
 49 allocated memory by gem
 50 -----------------------------------
 51   27243140  coverband/lib
 52   26388540  redis-4.1.0
 53     181300  other
```

Performance has also improved on the new HashRedisStore. Storing the full set of files is still slower but storing the subset is around 3X faster.
```
runs benchmarks on reporting large sets of files to redis
Warming up --------------------------------------
   store_reports_all     1.000  i/100ms
Calculating -------------------------------------
   store_reports_all      3.944  (± 0.0%) i/s -     60.000  in  15.247118s
Warming up --------------------------------------
store_reports_subset    11.000  i/100ms
Calculating -------------------------------------
store_reports_subset    116.872  (± 6.0%) i/s -      2.332k in  20.028811s
```

**Comment by danmayer** [comment]
Nice, that is an interesting idea using lua... that is cool.

as for the sparseness, my guess is that 0 is more comment than the nil records but we would have to only pick one to go sparse on. Nil makes more sense logically, but might be less valuable. 

I will try to look a bit closer, but I guess we might want to test this on some more realistic apps to get perf data.

**Comment by danmayer** [comment]
This should be resolved in the latest release candidate.
